loader/db.go

```diff
@@ -34,213 +36,151 @@ import (
-				err = tmysql.NewErr(uint16(errCode))
-				ctx.L().Warn("executeSQLCustomRetry failed", zap.String("failpoint", "LoadExecCreateTableFailed"), zap.Error(err))
-			}
-		})
-
-		if err != nil {
-			tidbExecutionErrorCounter.WithLabelValues(conn.cfg.Name).Inc()
-			if isRetryableFn(err) {
-				continue
-			}
-			return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, strings.Join(sqls, ";"))
-		}
-
-		// update metrics
-		cost := time.Since(startTime)
-		txnHistogram.WithLabelValues(conn.cfg.Name).Observe(cost.Seconds())
-		if cost > 1 {
-			ctx.L().Warn("transaction execute successfully", zap.Duration("cost time", cost))
+	sqls := make([]baseconn.SQL, 0, len(queries))
+	for i, query := range queries {
+		if i >= len(args) {
+			sqls = append(sqls, baseconn.SQL{Query: query, Args: []interface{}{}})
+		} else {
+			sqls = append(sqls, baseconn.SQL{Query: query, Args: args[i]})
 		}
-
-		return nil
 	}
 
-	return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, strings.Join(sqls, ";"))
+	return conn.executeSQLCustomRetry(ctx, sqls, retry.IsLoaderRetryableError)
 }
 
-func executeSQLImp(ctx *tcontext.Context, db *sql.DB, sqls []string) error {
-	var (
-		err error
-		txn *sql.Tx
-		res sql.Result
-	)
-
-	txn, err = db.Begin()
-	if err != nil {
-		ctx.L().Error("fail to begin a transaction", zap.String("sqls", fmt.Sprintf("%-.200v", sqls)), zap.Error(err))
-		return terror.DBErrorAdapt(err, terror.ErrDBDriverError)
+func (conn *Conn) executeSQLCustomRetry(ctx *tcontext.Context, sqls []baseconn.SQL, retryFn func(err error) bool) error {
+	if conn == nil || conn.baseConn == nil {
+		return terror.ErrDBUnExpect.Generate("database connection not valid")
 	}
 
-	for i := range sqls {
-		ctx.L().Debug("execute statement", zap.String("sqls", fmt.Sprintf("%-.200v", sqls[i])))
-		res, err = txn.ExecContext(ctx.Context(), sqls[i])
-		if err != nil {
-			ctx.L().Warn("execute statement", zap.String("sqls", fmt.Sprintf("%-.200v", sqls[i])), log.ShortError(err))
-			rerr := txn.Rollback()
-			if rerr != nil {
-				ctx.L().Error("fail rollback", log.ShortError(rerr))
-			}
-			return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, sqls[i])
-		}
-		// check update checkpoint successful or not
-		if i == 2 {
-			row, err1 := res.RowsAffected()
-			if err1 != nil {
-				ctx.L().Warn("fail to get rows affected", zap.String("sqls", fmt.Sprintf("%-.200v", sqls[i])), log.ShortError(err1))
-				continue
+	params := retry.Params{
+		RetryCount:         10,
+		FirstRetryDuration: 2 * time.Second,
+		BackoffStrategy:    retry.LinearIncrease,
+		IsRetryableFn: func(retryTime int, err error) bool {
+			ctx.L().Debug("execute statement", zap.Int("retry", retryTime), zap.String("sqls", fmt.Sprintf("%-.200v", sqls)))
+			tidbExecutionErrorCounter.WithLabelValues(conn.cfg.Name).Inc()
+			if retryFn(err) {
+				return true
 			}
-			if row != 1 {
-				ctx.L().Warn("update checkpoint", zap.Int64("affected rows", row))
+			return false
```

Comment:
Is `return retryFn(err)` simpler?

Reply:
done in 
 https://github.com/pingcap/dm/pull/236/commits/59dacaa6bceb5fd19f6360808d5875fdb48a01c5#diff-5f5bb90e4d87dad0cb25c377bb10679bR121